source/lib/src/cuda/tabulate.cu

```diff
@@ -14,466 +10,231 @@
 #define FULL_MASK 0xffffffff
 
 template <typename FPTYPE>
-__forceinline__ 
-__device__
-void locate_xx(const FPTYPE& lower, const FPTYPE& upper,  const FPTYPE& max, const FPTYPE& stride0, const FPTYPE& stride1, FPTYPE& xx, int& table_idx) {
-    if (xx < lower) {
-        table_idx = 0;
-        xx = 0;
-    }
-    else if (xx < upper) {
-        table_idx = (int)((xx - lower) / stride0);
-        xx -= (table_idx * stride0 + lower);
-    }
-    else if (xx < max) {
-        int first_stride = int((upper - lower) / stride0);
-        table_idx = first_stride + (int)((xx - upper) / stride1);
-        xx -= ((table_idx - first_stride) * stride1 + upper);
-    }
-    else {
-        table_idx = int((upper - lower) / stride0) + (int)((max - upper) / stride1) - 1;
-        xx = 0;
-    }
+__forceinline__ __device__
+void locate_xx(
+    FPTYPE& xx, 
+    int& table_idx,
+    const FPTYPE& lower, 
+    const FPTYPE& upper,  
+    const FPTYPE& max, 
+    const FPTYPE& stride0, 
+    const FPTYPE& stride1)
+{
+  if (xx < lower) {
+    table_idx = 0;
+    xx = 0;
+  }
+  else if (xx < upper) {
+    table_idx = (int)((xx - lower) / stride0);
+    xx -= (table_idx * stride0 + lower);
+  }
+  else if (xx < max) {
+    int first_stride = int((upper - lower) / stride0);
+    table_idx = first_stride + (int)((xx - upper) / stride1);
+    xx -= ((table_idx - first_stride) * stride1 + upper);
+  }
+  else {
+    table_idx = int((upper - lower) / stride0) + (int)((max - upper) / stride1) - 1;
+    xx = 0;
+  }
 }
 
 template <typename FPTYPE>
-__forceinline__ 
-__device__ 
-FPTYPE dot(FPTYPE ll[4], FPTYPE rr[4]) {
-    return ll[0] * rr[0] + ll[1] * rr[1] + ll[2] * rr[2] + ll[3] * rr[3];
+__forceinline__ __device__ 
+FPTYPE dot(
```

Comment:
Place all `dot` in `utilities.h` ?

Reply:
It's a device function, nvcc compiler is needed, so better to leave it here.